File: backend/models/opportunity.py
import json

from .trigger import Trigger
from .event import Event

class Opportunity:
    def __init__(
        self,
        trigger: Trigger,
        bet_one_bookmaker: str, 
        bet_one_bookmaker_key: str, 
        bet_one_price: float, 
        bet_two_bookmaker: str, 
        bet_two_bookmaker_key: str, 
        bet_two_price: float, 
        event: Event, 
        stake: float
    ) -> None:
        self.trigger = trigger
        self.bet_one_bookmaker = bet_one_bookmaker
        self.bet_one_bookmaker_key = bet_one_bookmaker_key
        self.bet_one_price = bet_one_price
        self.bet_two_bookmaker = bet_two_bookmaker
        self.bet_two_bookmaker_key = bet_two_bookmaker_key
        self.bet_two_price = bet_two_price
        self.event = event
        self.stake = stake
        self.round = 0

    def compute_round(self):
        bet_one_percent = 100 / self.bet_one_price
        bet_two_percent = 100 / self.bet_two_price
        self.round = bet_one_percent + bet_two_percent

    def compute_stakes(self):
        self.bet_one_stake = ((100 / self.bet_one_price) / self.round) * self.stake
        self.bet_two_stake = ((100 / self.bet_two_price) / self.round) * self.stake
    
    def compute_returns(self):
        self.bet_one_return = self.bet_one_stake * self.bet_one_price
        self.bet_two_return = self.bet_two_stake * self.bet_two_price

    def compute_profits(self):
        self.bet_one_profit = self.bet_one_return - self.stake
        self.bet_two_profit = self.bet_two_return - self.stake

    def toJson(self):
        return json.dumps(self, default = lambda o: o.__dict__, sort_keys = True, indent = 4)

File: frontend/src/components/molecules/opportunity.js
import Card from 'react-bootstrap/Card';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import Moment from 'react-moment';

import Logo from '../atoms/logo/logo';

import './opportunity.css';

function american(decimal) {
    if (decimal >= 2) {
        return '+' + Math.ceil((decimal - 1) * 100).toString()
    } else {
        return '-' + (-1 * Math.ceil(-100 / (decimal - 1))).toString()
    }
}

function Opportunity({ opportunity }) {
    return(
        <Card>
            <Card.Body>
                <Row>
                     <Col xs={4} md={2}>
                        <h6>Time</h6>
                        <p><Moment format='MM/DD/YYYY HH:mm'>{opportunity.event.time}</Moment></p>
                    </Col>
                    <Col xs={4} md={2}>
                        <h6>Competition</h6>
                        <p>{opportunity.event.sport_title}</p>
                    </Col>
                    <Col xs={4} md={2}>
                        <h6>Event</h6>
                        <p>{opportunity.event.home_team} vs {opportunity.event.away_team}</p>
                    </Col>
                    <Col xs={4} md={2}>
                        <h6>Market</h6>
                        <p>{opportunity.trigger.type}</p>
                    </Col>
                    <Col xs={4} md={2}>
                        <h6>Bet 1</h6>
                        <Logo bookmaker_key={opportunity.bet_one_bookmaker_key} />
                        <h6>{opportunity.trigger.bet_one_dir} : {opportunity.bet_one_bookmaker}</h6>
                        <p>Odds: {american(opportunity.bet_one_price)}</p>
                        <p>Stake: ${opportunity.bet_one_stake.toFixed(2)}</p>
                        <p>Profit: ${opportunity.bet_one_profit.toFixed(2)}</p>
                    </Col>
                    <Col xs={4} md={2}>
                        <h6>Bet 2</h6>
                        <Logo bookmaker_key={opportunity.bet_two_bookmaker_key} />
                        <h6>{opportunity.trigger.bet_two_dir} : {opportunity.bet_two_bookmaker}</h6>
                        <p>Odds: {american(opportunity.bet_two_price)}</p>
                        <p>Stake: ${opportunity.bet_two_stake.toFixed(2)}</p>
                        <p>Profit: ${opportunity.bet_two_profit.toFixed(2)}</p>
                    </Col>
                </Row>
            </Card.Body>
        </Card>
    )
}

export default Opportunity;